# Data Science Part Time Course

## Week 5 - Lesson 2 - Lab: Databases with Python

In this Lab the goal is connect to a database (created form a local file), read the data into Python and interact with it.

## 1. Data import

Connect to the file 'assets/datasets/enron.db' using one of these methods:

- sqlite3 python package
- pandas.read_sql
- SQLite Manager Firefox extension

Take a look at the database and query the master table. How many Tables are there in the db?

> Answer:
There are 3 tables:
- MessageBase
- RecipientBase
- EmployeeBase

In [1]:
import sqlite3
conn = sqlite3.connect('enron.db') 
c = conn.cursor()
c.execute("SELECT name FROM sqlite_master WHERE type='table';").fetchall()

[('MessageBase',), ('RecipientBase',), ('EmployeeBase',)]

Query the `sqlite_master` table to retrieve the **schema** of the `EmployeeBase` table.

1. What fields are there?
1. What's the type of each of them?

In [2]:
fields = c.execute("SELECT sql from sqlite_master WHERE type='table' and name='MessageBase';").fetchall()
print (''.join(fields[0]))

fields = c.execute("SELECT sql from sqlite_master WHERE type='table' and name='RecipientBase';").fetchall()
print (''.join(fields[0]))

fields = c.execute("SELECT sql from sqlite_master WHERE type='table' and name='EmployeeBase';").fetchall()
print (''.join(fields[0]))

CREATE TABLE MessageBase (
    mid INTEGER,
    filename TEXT,
    unix_time INTEGER,
    subject TEXT,
    from_eid INTEGER,
    
    PRIMARY KEY(mid ASC),
    FOREIGN KEY(from_eid) REFERENCES Employee(eid)
)
CREATE TABLE RecipientBase (
    mid INTEGER,
    rno INTEGER,
    to_eid INTEGER,
    
    PRIMARY KEY(mid ASC, rno ASC)
    FOREIGN KEY(mid) REFERENCES Message(mid)
    FOREIGN KEY(to_eid) REFERENCES Employee(eid)
)
CREATE TABLE EmployeeBase (
                  [eid] INTEGER,
  [name] TEXT,
  [department] TEXT,
  [longdepartment] TEXT,
  [title] TEXT,
  [gender] TEXT,
  [seniority] TEXT
                  
                  )


1. Print the first 5 rows of EmployeeBase table
1. Print the first 5 rows of MessageBase table
1. Print the first 5 rows of RecipientBase table

**Hint**  use `SELECT` and `LIMIT`.

In [3]:
results = c.execute("SELECT * FROM EmployeeBase LIMIT 5;").fetchall()
for row in results:
    print (row)

(1, 'John Arnold', 'Forestry', 'ENA Gas Financial', 'VP Trading', 'Male', 'Senior')
(2, 'Harry Arora', 'Forestry', 'ENA East Power', 'VP Trading', 'Male', 'Senior')
(3, 'Robert Badeer', 'Forestry', 'ENA West Power', 'Mgr Trading', 'Male', 'Junior')
(4, 'Susan Bailey', 'Legal', 'ENA Legal', 'Specialist Legal', 'Female', 'Junior')
(5, 'Eric Bass', 'Forestry', 'ENA Gas Texas', 'Trader', 'Male', 'Junior')


In [4]:
results = c.execute("SELECT * FROM MessageBase LIMIT 5;").fetchall()
for row in results:
    print (row)

(1, 'taylor-m/sent/11', 910930020, 'Cd$ CME letter', 138)
(2, 'taylor-m/sent/17', 911459940, 'Indemnification', 138)
(3, 'taylor-m/sent/18', 911463840, 'Re: Indemnification', 138)
(4, 'taylor-m/sent/23', 911874180, 'Re: Coral Energy, L.P.', 138)
(5, 'taylor-m/sent/27', 912396120, 'Bankruptcy Code revisions', 138)


In [5]:
results = c.execute("SELECT * FROM RecipientBase LIMIT 10;").fetchall()
for row in results:
    print (row)
    
# The first field is message id, the second is recipient number, and the third is the id of the recipient.
# mid, rno, to_eid

(1, 1, 59)
(2, 1, 15)
(3, 1, 15)
(4, 1, 109)
(4, 2, 49)
(4, 3, 120)
(4, 4, 59)
(5, 1, 45)
(5, 2, 53)
(6, 1, 113)


Now try other SQL statements on the local database, such as SELECT .... FROM ... WHERE ....

For example, print the records of EmployeeBase where the Gender is male 

In [6]:
results = c.execute("SELECT * FROM EmployeeBase WHERE gender='Male'").fetchall()
for row in results:
    print(row)

(1, 'John Arnold', 'Forestry', 'ENA Gas Financial', 'VP Trading', 'Male', 'Senior')
(2, 'Harry Arora', 'Forestry', 'ENA East Power', 'VP Trading', 'Male', 'Senior')
(3, 'Robert Badeer', 'Forestry', 'ENA West Power', 'Mgr Trading', 'Male', 'Junior')
(5, 'Eric Bass', 'Forestry', 'ENA Gas Texas', 'Trader', 'Male', 'Junior')
(6, 'Don Baughman Jr.', 'Forestry', 'ENA East Power', 'Mgr Trading', 'Male', 'Junior')
(8, 'Robert Benson', 'Forestry', 'ENA East Power', 'Dir Trading', 'Male', 'Senior')
(11, 'Rick Buy', 'Forestry', 'Enron', 'VP & CRO', 'Male', 'Senior')
(12, 'Larry Campbell', 'Other', 'ETS', 'Specialist', 'Male', 'Junior')
(13, 'Larry F. Campbell', 'Other', 'ENA East Power', 'Specialist', 'Male', 'Junior')
(14, 'Mike Carson', 'Forestry', 'ENA East Power', 'Mgr Trading', 'Male', 'Junior')
(18, 'Sean Crandall', 'Forestry', 'ENA West Power', 'Dir Trading', 'Male', 'Senior')
(19, 'Martin Cuilla', 'Forestry', 'ENA Gas Central', 'Mgr Trading', 'Male', 'Junior')
(20, 'Jeff Dasovich', 'Legal

#### Import each of the 3 tables to a Pandas Dataframes

In [7]:
import pandas as pd
employees = pd.read_sql("SELECT * FROM EmployeeBase;", conn)
recipients = pd.read_sql("SELECT * FROM RecipientBase;", conn)
messages = pd.read_sql("SELECT * FROM MessageBase;", conn)

In [8]:
recipients.head(10)

mid  rno  to_eid
0    1    1      59
1    2    1      15
2    3    1      15
3    4    1     109
4    4    2      49
5    4    3     120
6    4    4      59
7    5    1      45
8    5    2      53
9    6    1     113

## 2. Data Exploration

Use the 3 dataframes to answer the following questions:

How many employees are there in the company? 156

In [9]:
employees.describe()

eid
count  156.000000
mean    78.500000
std     45.177428
min      1.000000
25%     39.750000
50%     78.500000
75%    117.250000
max    156.000000

How many messages are there in the database?

In [10]:
recipients.count()

mid       38388
rno       38388
to_eid    38388
dtype: int64

Some messages are sent to more than one recipient. Group the messages by message_id and count the number of recepients. Then look at the distribution of recepient numbers.
    - How many messages have only one recipient?
    - How many messages have >= 5 recipients?
    - What's the highest number of recipients?
    - Who sent the message with the highest number of recipients?

In [11]:
singlerec = recipients.rno == 1
recipients[singlerec].count()

mid       21635
rno       21635
to_eid    21635
dtype: int64

In [12]:
overfive = recipients.rno >= 5
recipients[overfive].count()

mid       4162
rno       4162
to_eid    4162
dtype: int64

In [13]:
most = recipients.rno == 57
print(recipients[most])

         mid  rno  to_eid
21179  12116   57      51
21361  12151   57      51


In [14]:
print(messages[messages.mid == 12116])
print(messages[messages.mid == 12151])

         mid                      filename  unix_time subject  from_eid
12115  12116  baughman-d/all_documents/398  990510780                67
         mid                 filename  unix_time subject  from_eid
12150  12151  baughman-d/ect_admin/22  990546780                67


* 38,388 messages in the recipient database, and 21,635 were sent to only a single recipient. 4,162 were sent to over five recipients. 
* The highest number of recipients is 57.
* Two messages were sent to 57 people, messages with ID 12116 and 12151. Both of these messages were sent by Baughman D.


Plot the distribution of recepient numbers using Bokeh.

In [15]:
import numpy as np
import scipy.special
from bokeh.layouts import gridplot
from bokeh.plotting import figure, output_notebook, show

p1 = figure(title="Distribution of Recipients",tools="save")
hist, edges = np.histogram(recipients.rno, density=False, bins=10)
x = np.arange(0,57)
p1.quad(top=hist, bottom=0, left=edges[:10], right=edges[1:], fill_color="#008080")
p1.legend.location = "center_right"
p1.xaxis.axis_label = 'No of Recipients'
p1.yaxis.axis_label = 'Messages'
output_notebook()
show(p1)

Loading BokehJS ...

## 3. Data Merging

Use the pandas merge function to combine the information in the 3 dataframes to answer the following questions:

1. Are there more Men or Women employees?
- How is gender distributed across departments?
- Who is sending more emails? Men or Women?
- What's the average number of emails sent by each gender?
- Are there more Juniors or Seniors?
- Who is sending more emails? Juniors or Seniors?
- Which department is sending more emails? How does that relate with the number of employees in the department?
- Who are the top 3 senders of emails? (people who sent out the most emails)

In [16]:
rm = pd.merge(recipients,messages,on="mid")
rm

mid  rno  to_eid                                      filename  \
0          1    1      59                              taylor-m/sent/11   
1          2    1      15                              taylor-m/sent/17   
2          3    1      15                              taylor-m/sent/18   
3          4    1     109                              taylor-m/sent/23   
4          4    2      49                              taylor-m/sent/23   
5          4    3     120                              taylor-m/sent/23   
6          4    4      59                              taylor-m/sent/23   
7          5    1      45                              taylor-m/sent/27   
8          5    2      53                              taylor-m/sent/27   
9          6    1     113                              taylor-m/sent/31   
10         7    1      59                              taylor-m/sent/33   
11         8    1      59                              taylor-m/sent/40   
12         9    1      49                              taylor-m/sent/41   
13        10    1      59                              taylor-m/sent/44   
14        11    1      59                              taylor-m/sent/45   
15        12    1      49                              taylor-m/sent/52   
16        12    2     120                              taylor-m/sent/52   
17        12    3      59                              taylor-m/sent/52   
18        13    1      59                              taylor-m/sent/57   
19        14    1      59                              taylor-m/sent/62   
20        15    1     138                     taylor-m/all_documents/73   
21        16    1      58                              taylor-m/sent/71   
22        16    2      26                              taylor-m/sent/71   
23        16    3      45                              taylor-m/sent/71   
24        16    4      53                              taylor-m/sent/71   
25        17    1     138                   taylor-m/all_documents/8545   
26        17    2     113                   taylor-m/all_documents/8545   
27        17    3      53                   taylor-m/all_documents/8545   
28        18    1      53                              taylor-m/sent/77   
29        19    1      59                              taylor-m/sent/78   
...      ...  ...     ...                                           ...   
38358  21627    1      39  germany-c/bankrupt/transport/national_fuel/4   
38359  21628    1      55                          germany-c/personal/3   
38360  21628    2      76                          germany-c/personal/3   
38361  21628    3      62                          germany-c/personal/3   
38362  21628    4     112                          germany-c/personal/3   
38363  21628    5      69                          germany-c/personal/3   
38364  21628    6     140                          germany-c/personal/3   
38365  21628    7      79                          germany-c/personal/3   
38366  21628    8     156                          germany-c/personal/3   
38367  21628    9      43                          germany-c/personal/3   
38368  21628   10     152                          germany-c/personal/3   
38369  21628   11      39                          germany-c/personal/3   
38370  21628   12      66                          germany-c/personal/3   
38371  21628   13      30                          germany-c/personal/3   
38372  21628   14     128                          germany-c/personal/3   
38373  21628   15      36                          germany-c/personal/3   
38374  21628   16       3                          germany-c/personal/3   
38375  21629    1      39                            germany-c/inbox/66   
38376  21630    1      39                          germany-c/calendar/5   
38377  21631    1     120                            germany-c/inbox/49   
38378  21632    1      57                       germany-c/sent_items/74   
38379  21633    1      39         

In [17]:
rme = pd.merge(rm,employees,left_on="from_eid",right_on="eid")
rme

mid  rno  to_eid                            filename   unix_time  \
0          1    1      59                    taylor-m/sent/11   910930020   
1          2    1      15                    taylor-m/sent/17   911459940   
2          3    1      15                    taylor-m/sent/18   911463840   
3          4    1     109                    taylor-m/sent/23   911874180   
4          4    2      49                    taylor-m/sent/23   911874180   
5          4    3     120                    taylor-m/sent/23   911874180   
6          4    4      59                    taylor-m/sent/23   911874180   
7          5    1      45                    taylor-m/sent/27   912396120   
8          5    2      53                    taylor-m/sent/27   912396120   
9          6    1     113                    taylor-m/sent/31   912570420   
10         7    1      59                    taylor-m/sent/33   912576240   
11         8    1      59                    taylor-m/sent/40   912685080   
12         9    1      49                    taylor-m/sent/41   912734100   
13        10    1      59                    taylor-m/sent/44   913166040   
14        11    1      59                    taylor-m/sent/45   913166220   
15        12    1      49                    taylor-m/sent/52   913770180   
16        12    2     120                    taylor-m/sent/52   913770180   
17        12    3      59                    taylor-m/sent/52   913770180   
18        13    1      59                    taylor-m/sent/57   913797300   
19        14    1      59                    taylor-m/sent/62   913950120   
20        16    1      58                    taylor-m/sent/71   914298240   
21        16    2      26                    taylor-m/sent/71   914298240   
22        16    3      45                    taylor-m/sent/71   914298240   
23        16    4      53                    taylor-m/sent/71   914298240   
24        18    1      53                    taylor-m/sent/77   915416460   
25        19    1      59                    taylor-m/sent/78   915423060   
26        20    1      45                    taylor-m/sent/82   915698520   
27        21    1      59                    taylor-m/sent/83   915698760   
28        22    1      49                    taylor-m/sent/84   915699540   
29        23    1     130                    taylor-m/sent/86   915766680   
...      ...  ...     ...                                 ...         ...   
38358  21503    5     115              lokey-t/sent_items/157  1016625760   
38359  21556    1     101              lokey-t/sent_items/159  1017059229   
38360  16345    1     133            ring-r/deleted_items/162  1002786683   
38361  18541    1     131  steffes-j/security___collateral/11  1005751023   
38362  20454    1      20                ring-r/sent_items/33  1011268223   
38363  20454    2     131                ring-r/sent_items/33  1011268223   
38364  20455    1      20                ring-r/sent_items/34  1011268442   
38365  20455    2     131                ring-r/sent_items/34  1011268442   
38366  20499    1     131                ring-r/sent_items/35  1011347459   
38367  16751    1      30             ermis-f/deleted_items/5  1003402071   
38368  16796    1      30                   ermis-f/inbox/715  1003471662   
38369  18897    1      43         grigsby-m/deleted_items/375  1006254872   
38370  21119    1     118           scott-s/deleted_items/214  1013510358   
38371  21128    1     118           scott-s/deleted_items/212  1013516057   
38372  21129    1     118           scott-s/deleted_items/211  1013516117   
38373  21132    1     118           scott-s/deleted_items/210  1013516278   
38374  21134    1     118           scott-s/deleted_items/209  1013516506   
38375  21211    1     118           scott-s/deleted_items/220  1014102235   
38376  21214    1     118           scott-s/deleted_items/219  1014102942   
38377  17356    1      98                davis-d/sent_items/5  1004097708   
38378

In [18]:
#Are there more Men or Women employees?
employees['gender'].value_counts()
#There are more men than women.

Male      113
Female     43
Name: gender, dtype: int64

In [19]:
#How is gender distributed across departments?
employees.groupby(["department", "gender"]).size()

department  gender
Forestry    Female    10
            Male      50
Legal       Female    13
            Male      12
Other       Female    20
            Male      51
dtype: int64

In [20]:
#Who is sending more emails? Men or Women?
#What's the average number of emails sent by each gender?
#More emails are sent by men, but there are more male employees.
employees['gender'].value_counts()

Male      113
Female     43
Name: gender, dtype: int64

In [21]:
rme['gender'].value_counts()

Male      22798
Female    15590
Name: gender, dtype: int64

In [22]:
msends=22798/113
fsends=15590/43
print("Men sent ",msends,"Women sent ",fsends)
#Women are sending more emails than men.

Men sent  201.75221238938053 Women sent  362.5581395348837


In [23]:
#Are there more Juniors or Seniors?
employees.groupby(['seniority']).size()

seniority
Junior    82
Senior    74
dtype: int64

In [24]:
#Who is sending more emails? Juniors or Seniors?
rme.groupby(['seniority']).size()

seniority
Junior    15898
Senior    22490
dtype: int64

In [25]:
jsendrate = 15898/82
ssendrate = 22490/74
print(jsendrate,ssendrate)
# Seniors are sending more emails.

193.8780487804878 303.9189189189189


In [26]:
#Which department is sending more emails? How does that relate with the number of employees in the department?
employees['department'].value_counts()

Other       71
Forestry    60
Legal       25
Name: department, dtype: int64

In [27]:
rme['department'].value_counts()

Legal       19292
Other       11984
Forestry     7112
Name: department, dtype: int64

In [28]:
fsendrate = 7112 / 25
lsendrate = 19292 / 25
osendrate = 11984 / 71
print(fsendrate,lsendrate,osendrate)
#The Legal department sends the most emails, but has the fewest employees.

284.48 771.68 168.7887323943662


In [29]:
#Who are the top 3 senders of emails? (people who sent out the most emails)
rme.sort_values(by='name') 
rme['name'].value_counts()
#Jeff Dasovich        3844
#Tana Jones           2858
#James D. Steffes     1666

Jeff Dasovich        3844
Tana Jones           2858
James D. Steffes     1666
Sara Shackleton      1515
Mike Grigsby         1204
Bill Williams III    1045
Carol St Clair       1019
Mark E. Taylor        918
Mary Hain             905
Stephanie Panus       822
Susan Scott           797
John J. Lavorato      775
Louise Kitchen        723
Drew Fossum           676
Chris Germany         628
Kimberly Watson       627
Steven J. Kean        613
Richard Shapiro       522
David W. Delainey     518
Debra Perlingiere     501
Gerald Nemec          480
Phillip M. Love       436
Marie Heard           434
Eric Bass             428
Michelle Lokay        416
John Arnold           413
Shelley Corman        396
Rod Hayslett          396
Mark E. Haedicke      393
Matthew Lenhart       383
                     ... 
Frank Ermis            15
Robert Badeer          15
Kevin Ruscitti         15
Judy Townsend          15
Mike Carson            14
Eric Linder            13
Joe Stepenovitch       12
Lisa Gang   